test/test_distributions.py

```diff
@@ -1055,5 +1068,107 @@ def test_laplace_shape_tensor_params(self):
         self.assertRaises(ValueError, laplace.log_prob, self.tensor_sample_2)
 
 
+class TestNumericalStability(TestCase):
+    def _test_pdf_score(self,
+                        dist_class,
+                        x,
+                        expected_value,
+                        probs=None,
+                        logits=None,
+                        expected_gradient=None,
+                        prec=1e-5):
+        if probs is not None:
+            p = Variable(probs, requires_grad=True)
+            dist = dist_class(p)
+        else:
+            p = Variable(logits, requires_grad=True)
+            dist = dist_class(logits=p)
+        log_pdf = dist.log_prob(Variable(x))
+        log_pdf.sum().backward()
+        self.assertEqual(log_pdf.data,
+                         expected_value,
+                         prec=prec,
+                         message='Failed for tensor type: {}. Expected = {}, Actual = {}'
+                         .format(type(x), expected_value, log_pdf.data))
+        if expected_gradient is not None:
+            self.assertEqual(p.grad.data,
+                             expected_gradient,
+                             prec=prec,
+                             message='Failed for tensor type: {}. Expected = {}, Actual = {}'
+                             .format(type(x), expected_gradient, p.grad.data))
+
+    def test_bernoulli_gradient(self):
+        for tensor_type in [torch.FloatTensor, torch.DoubleTensor]:
+            self._test_pdf_score(dist_class=Bernoulli,
+                                 probs=tensor_type([0]),
+                                 x=tensor_type([0]),
+                                 expected_value=tensor_type([0]),
+                                 expected_gradient=tensor_type([0]))
+
+            self._test_pdf_score(dist_class=Bernoulli,
+                                 probs=tensor_type([0]),
+                                 x=tensor_type([1]),
+                                 expected_value=tensor_type([_get_clamping_buffer(tensor_type([]))]).log(),
+                                 expected_gradient=tensor_type([0]))
+
+            self._test_pdf_score(dist_class=Bernoulli,
+                                 probs=tensor_type([1e-4]),
+                                 x=tensor_type([1]),
+                                 expected_value=tensor_type([math.log(1e-4)]),
+                                 expected_gradient=tensor_type([10000]))
+
+            # Lower precision due to:
+            # >>> 1 / (1 - torch.FloatTensor([0.9999]))
+            # 9998.3408
```

Comment:
@apaszke - I was a bit surprised by this. Any ideas why this does not return `10,000`, as you might expect?

Reply:
Because 1 - 0.9999 = 9.999999999998899e-05 (that's in double precision, but pretty much the same thing happens in fp32).

Reply:
Thanks. What I found interesting for `FloatTensor` (`DoubleTensor` returns expected result to first five decimal places at least) was that the result is less than `10,000` for `0.9999` and beyond. I am probably missing some floating point arithmetic details here! 

e.g.
```
>>> 1/(1-torch.FloatTensor([0.99]))
 100.0001
[torch.FloatTensor of size 1]

>>> 1/(1-torch.FloatTensor([0.999]))
 1000.0129
[torch.FloatTensor of size 1]

>>> 1/(1-torch.FloatTensor([0.9999]))
 9998.3408
[torch.FloatTensor of size 1]

```